internal/resources/cloudprovider/models.go

```diff
@@ -41,30 +42,82 @@ type awsCWScrapeJobTFDataSourceModel struct {
 	// TODO(tristan): if the grafana provider is updated to use the Terraform v6 plugin protocol,
 	// we can consider adding additional support to use Set Nested Attributes, instead of Blocks.
 	// See https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes#nested-attribute-types
-	Services         []awsCWScrapeJobServiceTFModel         `tfsdk:"service"`
-	CustomNamespaces []awsCWScrapeJobCustomNamespaceTFModel `tfsdk:"custom_namespace"`
+	Services         types.List `tfsdk:"service"`
+	CustomNamespaces types.List `tfsdk:"custom_namespace"`
 }
+
 type awsCWScrapeJobServiceTFModel struct {
-	Name                        types.String                     `tfsdk:"name"`
-	Metrics                     []awsCWScrapeJobMetricTFModel    `tfsdk:"metric"`
-	ScrapeIntervalSeconds       types.Int64                      `tfsdk:"scrape_interval_seconds"`
-	ResourceDiscoveryTagFilters []awsCWScrapeJobTagFilterTFModel `tfsdk:"resource_discovery_tag_filter"`
-	TagsToAddToMetrics          types.Set                        `tfsdk:"tags_to_add_to_metrics"`
+	Name                        types.String `tfsdk:"name"`
+	Metrics                     types.List   `tfsdk:"metric"`
+	ScrapeIntervalSeconds       types.Int64  `tfsdk:"scrape_interval_seconds"`
+	ResourceDiscoveryTagFilters types.List   `tfsdk:"resource_discovery_tag_filter"`
+	TagsToAddToMetrics          types.Set    `tfsdk:"tags_to_add_to_metrics"`
+}
+
+func (m awsCWScrapeJobServiceTFModel) attrTypes() map[string]attr.Type {
```

Comment:
what are this for?

Reply:
this is for plugging into Terraform schema validation in the conversion functions where the inner type of the List is the generic TF Object type. The map lets TF validate conformity with the Go-type-specific schema, if that makes sense.

Reply:
Yep, makes sense, thanks for explaining!